<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Mapping mFISH data to reference data set • mfishtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Mapping mFISH data to reference data set">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mfishtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/inhibitory_marker_mapping.html">Mapping mFISH data to reference data set</a>
    </li>
    <li>
      <a href="../articles/inhibitory_marker_selection.html">Marker panel generation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Mapping mFISH data to reference data set</h1>
            
      

      <div class="hidden name"><code>inhibitory_marker_mapping.Rmd</code></div>

    </div>

    
    
<p>This code reads in all of the data for an example mouse SST mFISH
experiment and compares it against a reference FACs data set. <strong>In
this case we have very limited prior knowledge, so I am open to
suggestions for how to check whether the results look
reasonable.</strong>.</p>
<div class="section level3">
<h3 id="workspace-set-up">Workspace set-up<a class="anchor" aria-label="anchor" href="#workspace-set-up"></a>
</h3>
<p>Install the necessary packages. In this case we are using data from
<code>tasic2016data</code> and plotting functions from
<code>scrattch.vis</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"AllenInstitute/tasic2016data"</span><span class="op">)</span> <span class="co"># For our data example</span></span></code></pre></div>
<p>Load libraries.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mfishtools</span><span class="op">)</span>    <span class="co"># This library!</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">matrixStats</a></span><span class="op">)</span>   <span class="co"># For rowMedians function, which is fast</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/talgalili/gplots" class="external-link">gplots</a></span><span class="op">)</span>        <span class="co"># For some of the plots</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>       <span class="co"># For other plots</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tasic2016data</span><span class="op">)</span> <span class="co"># For the data</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>  <span class="co"># IMPORTANT</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Libraries loaded."</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Libraries loaded."</span></span></code></pre>
<p>Read in the reference data (in this case we will use the Tasic 2016
data, which includes ~1800 cells from mouse primary visual cortex).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="va">tasic_2016_anno</span></span>
<span><span class="co">#counts      &lt;- tasic_2016_counts  # uncomment if using CPM below</span></span>
<span><span class="va">rpkm</span>        <span class="op">&lt;-</span> <span class="va">tasic_2016_rpkm</span></span>
<span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="va">annotations</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">rpkm</span><span class="op">)</span>,<span class="va">annotations</span><span class="op">$</span><span class="va">sample_name</span><span class="op">)</span>,<span class="op">]</span>  <span class="co"># Put them in the correct order</span></span></code></pre></div>
<p>Read in the mFISH data. An example data set is provided as part of
<code>mfishtools</code>, and was loaded with the library. In this case,
<code>fishData</code> is a matrix of gene expression levels (e.g., spot
counts within a cell) for a given cell, with genes as rows and cells as
columns. This is the same format as the RNA-seq data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">fishData</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1]  22 171</span></span></code></pre>
<p>The <code>metadata</code> variable is a data frame with some specific
requirements on column names.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "area"       "experiment" "layerData"  "x"          "y"</span></span></code></pre>
<p>The required column names are as follows (and other column names are
perfectly fine):<br>
- area = Area/volume of the cell (or just set to a constant)<br>
- experiment = Name of the experiment or experiments (or just set ot a
constant)<br>
- layerData = Numeric call for the layer. Not requred, but useful for
plotting an rotating x, y<br>
- x = X coordinate for cell (ideally this is the lateral
coordinate)<br>
- y = Y coordinate for cell (ideally this is the laminar coordinate)</p>
</div>
<div class="section level3">
<h3 id="data-preparations">Data preparations<a class="anchor" aria-label="anchor" href="#data-preparations"></a>
</h3>
<p>This analysis will only be looking at marker genes for GABAergic
neurons, so we need to only consider cells mapping to GABAergic types.
We also define some convenient cluster info variables here.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusterType</span> <span class="op">=</span> <span class="va">annotations</span><span class="op">$</span><span class="va">broad_type</span> </span>
<span><span class="va">includeClas</span> <span class="op">=</span> <span class="st">"GABA-ergic Neuron"</span>  <span class="co"># In this analysis, we are only considering interneurons</span></span>
<span><span class="va">excludeClas</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">clusterType</span>,<span class="va">includeClas</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">kpSamp</span>      <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">is.element</a></span><span class="op">(</span><span class="va">clusterType</span>,<span class="va">excludeClas</span><span class="op">)</span></span>
<span><span class="va">anno</span>        <span class="op">=</span> <span class="va">annotations</span><span class="op">[</span><span class="va">kpSamp</span>,<span class="op">]</span></span>
<span><span class="va">cl</span>          <span class="op">=</span> <span class="va">annotations</span><span class="op">$</span><span class="va">primary_type_label</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span>   <span class="op">=</span> <span class="va">annotations</span><span class="op">$</span><span class="va">sample_name</span></span>
<span><span class="va">kpClust</span>     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cl</span><span class="op">[</span><span class="va">kpSamp</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Convert the data to log2(rpkm). NOTE: we often use counts per million
of introns + exons when performing this analysis. Currently, we don’t
know which method produces more reliable markers. Alternative code for
calculating cpm is commented out below.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">normDat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">rpkm</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#sf      = colSums(counts)/10^6</span></span>
<span><span class="co">#cpms    = t(t(counts)/sf)</span></span>
<span><span class="co">#normDat = log2(cpms+1)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Data normalized!"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Data normalized!"</span></span></code></pre>
<p>Calculate proportions, means, and medians. These are all used later
for various reasons. One important thing to note is that we are using
<code>cl</code> here as a vector of cell type calls for the RNA-seq data
(with names corresponding to the column/sample names of the RNA-seq
sample data); however, if you wanted to map to a more coarse definition
of cell types, or to omit certain cell types, this would be the step to
do it. The columns of these summary values calculated here define the
cell types that mFISH data will be mapped against for the remainder of
the vignette.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exprThresh</span> <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="va">medianExpr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span>, <span class="va">cl</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/rowMedians.html">rowMedians</a></span><span class="op">(</span><span class="va">normDat</span><span class="op">[</span>,<span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">meanExpr</span>   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span>, <span class="va">cl</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">normDat</span><span class="op">[</span>,<span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">propExpr</span>   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span>, <span class="va">cl</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">normDat</span><span class="op">[</span>,<span class="va">x</span><span class="op">]</span><span class="op">&gt;</span><span class="va">exprThresh</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">medianExpr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">propExpr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">meanExpr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">normDat</span><span class="op">)</span>  </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Summary values calculated!"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Summary values calculated!"</span></span></code></pre>
<p>Consider only genes included in the mFISH experiment that are also
present in the RNA-seq reference data set (usually this is all of
them).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">useGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">fishData</span><span class="op">)</span>,<span class="va">genes</span><span class="op">)</span>  <span class="co"># Define genes to be used in the analysis</span></span>
<span><span class="va">fishDat</span>  <span class="op">&lt;-</span> <span class="va">fishData</span><span class="op">[</span><span class="va">useGenes</span>,<span class="op">]</span>    <span class="co"># Separate out the data from the metadata</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"mFISH data is ready!"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "mFISH data is ready!"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="map-the-mfish-data">Map the mFISH data!<a class="anchor" aria-label="anchor" href="#map-the-mfish-data"></a>
</h2>
<p>Do the mapping using the parameters defined below. <em>Currently
these are all manually selected and a bit of a guess. I am open to
suggestions for how to do this mapping in a more systematic way.</em>
The idea behind this method is to attach various filtering and scaling
strategies to the mFISH and RNA-seq data sets, and then use
correlation-based mapping to find the best fitting cell cluster. My
expectation (still to be tested) is that this strategy will be useful
for smaller gene panels, but that other more computational-based
strategies will be most effective for larger gene panels.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qprob</span>    <span class="op">&lt;-</span> <span class="fl">0.9</span>     <span class="co"># Parameter for scaling mFISH to FACS</span></span>
<span><span class="va">thresh</span>   <span class="op">&lt;-</span> <span class="fl">3</span>       <span class="co"># Set counts less than or equal to thresh to 0          1</span></span>
<span><span class="va">log2p1</span>   <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">x</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>  <span class="co"># log transform function</span></span>
<span><span class="va">binarize</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span>   <span class="co"># Should the data be binarized?</span></span>
<span><span class="va">weights</span>  <span class="op">&lt;-</span> <span class="cn">NULL</span>    <span class="co"># Integer weights.  SET TO NULL IF YOU DON'T KNOW WHAT YOU ARE DOING!</span></span>
<span><span class="co">#weights  &lt;- round(rowSums(fishDat)/min(rowSums(fishDat)))  # Here is how to weight roughly by average expression level</span></span>
<span> </span>
<span><span class="va">fishMouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fishScaleAndMap.html">fishScaleAndMap</a></span><span class="op">(</span>mapDat<span class="op">=</span><span class="va">fishDat</span>, refSummaryDat<span class="op">=</span><span class="va">medianExpr</span>, </span>
<span>  mappingFunction <span class="op">=</span> <span class="va">cellToClusterMapping_byCor</span>, transform <span class="op">=</span> <span class="va">log2p1</span>, noiselevel <span class="op">=</span> <span class="va">thresh</span>, </span>
<span>  genesToMap <span class="op">=</span> <span class="va">useGenes</span>, metadata <span class="op">=</span> <span class="va">metadata</span>, qprob<span class="op">=</span><span class="va">qprob</span>, binarize<span class="op">=</span><span class="va">binarize</span>,</span>
<span>  omitGenes <span class="op">=</span> <span class="cn">NULL</span>,integerWeights<span class="op">=</span><span class="va">weights</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Data is mapped."</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Data is mapped."</span></span></code></pre>
<div class="section level4">
<h4 id="view-the-mapping-results">View the mapping results<a class="anchor" aria-label="anchor" href="#view-the-mapping-results"></a>
</h4>
<p>First rotate the X and Y coordinate space to that layer 2 is parallel
to the X axis. This won’t be perfect since the tissue is not perfectly
linear, but it will make for easier viewing of the data. This requires
some meta-data variable which marks a subset of genes roughly along
align (e.g., cortical layer calls).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rotateAxis</span> <span class="op">&lt;-</span> <span class="va">fishMouse</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">layerData</span><span class="op">==</span><span class="fl">4</span></span>
<span><span class="va">flipVector</span> <span class="op">&lt;-</span> <span class="va">fishMouse</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">layerData</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">e</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">fishMouse</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">experiment</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">subset</span>  <span class="op">&lt;-</span> <span class="va">fishMouse</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">experiment</span><span class="op">==</span><span class="va">e</span></span>
<span>  <span class="va">fishMouse</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rotateXY.html">rotateXY</a></span><span class="op">(</span><span class="va">fishMouse</span>,<span class="va">rotateAxis</span>,<span class="va">flipVector</span>,<span class="va">subset</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now plot the location of all of the cells in the tissue section.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc</span>  <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">n</span>,<span class="va">...</span><span class="op">)</span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown"</span>,<span class="st">"pink"</span>,<span class="st">"orange"</span>,<span class="st">"turquoise"</span>,<span class="st">"blue"</span>,<span class="st">"green"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n</span><span class="op">]</span><span class="op">)</span>  <span class="co"># Standard colors without yellow</span></span>
<span><span class="va">lay</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">fishMouse</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">layerData</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotDistributions.html">plotDistributions</a></span><span class="op">(</span><span class="va">fishMouse</span>,group <span class="op">=</span> <span class="st">"experiment"</span>, xlab<span class="op">=</span><span class="st">"Mouse - All cells"</span>, ylab<span class="op">=</span><span class="st">"Layer"</span>,</span>
<span>                  colors<span class="op">=</span><span class="va">lay</span>, colormap<span class="op">=</span><span class="va">sc</span>,maxrow<span class="op">=</span><span class="fl">8</span>,cex<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="inhibitory_marker_mapping_files/figure-html/plot%20cell%20locations-1.png" width="576"></p>
<p>Show the cell distribution across all types in the tissue.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allClusts</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">medianExpr</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotDistributions.html">plotDistributions</a></span><span class="op">(</span><span class="va">fishMouse</span>,group <span class="op">=</span> <span class="st">"Class"</span>, groups <span class="op">=</span> <span class="va">allClusts</span>, colors <span class="op">=</span> <span class="va">lay</span>, pch<span class="op">=</span><span class="va">lay</span>, xlab<span class="op">=</span><span class="st">"Mouse"</span>, </span>
<span>                  ylab<span class="op">=</span><span class="st">"Layer"</span>,colormap<span class="op">=</span><span class="va">sc</span>,maxrow<span class="op">=</span><span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<p><img src="inhibitory_marker_mapping_files/figure-html/plot%20mFISH%20distributions-1.png" width="1920"></p>
<p>There are some broad observations that suggest we are not too far
off. First, nearly all of the cells map to inhibitory types as expected.
Second, many of the cell types have layer signatures, with SST/PVALB
types more likely to be in deep layers and other inhibitory types more
likely to be in upper layers.</p>
<p>Now let’s plot the heatmap to see how the data looks when clustered
along the tree like this. First, unscaled data, capped at 10 counts.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cap</span> <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="va">colorset</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkblue"</span>, <span class="st">"dodgerblue"</span>, <span class="st">"gray80"</span>, <span class="st">"orange"</span>, <span class="st">"orangered"</span><span class="op">)</span></span>
<span><span class="va">heat_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="va">colorset</span><span class="op">)</span><span class="op">(</span><span class="fl">1001</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">fishMouse</span>,main<span class="op">=</span><span class="st">"Mouse cells (unscaled, cap=10)"</span>,group<span class="op">=</span><span class="st">"Class"</span>,groups<span class="op">=</span><span class="va">allClusts</span>,capValue<span class="op">=</span><span class="va">cap</span>,</span>
<span>                             colormap<span class="op">=</span><span class="va">heat_colors</span>,rowsep<span class="op">=</span><span class="cn">NULL</span>,dendrogram<span class="op">=</span><span class="st">"row"</span>,Rowv <span class="op">=</span> <span class="cn">TRUE</span>,margins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>,<span class="fl">6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="inhibitory_marker_mapping_files/figure-html/Plot%20mFISH%20heatmap%20(unscaled)-1.png" width="2304"></p>
<p>Next, let’s plot the scaled heatmaps for comparison. This is the data
that is used in the mapping.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cap</span> <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="va">colorset</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkblue"</span>, <span class="st">"dodgerblue"</span>, <span class="st">"gray80"</span>, <span class="st">"orange"</span>, <span class="st">"orangered"</span><span class="op">)</span></span>
<span><span class="va">heat_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="va">colorset</span><span class="op">)</span><span class="op">(</span><span class="fl">1001</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">fishMouse</span>,main<span class="op">=</span><span class="st">"Mouse cells (scaled, cap=8)"</span>,group<span class="op">=</span><span class="st">"Class"</span>,groups<span class="op">=</span><span class="va">allClusts</span>,capValue<span class="op">=</span><span class="va">cap</span>,colormap<span class="op">=</span><span class="va">heat_colors</span>,</span>
<span>                             rowsep<span class="op">=</span><span class="cn">NULL</span>,dendrogram<span class="op">=</span><span class="st">"row"</span>,Rowv <span class="op">=</span> <span class="cn">TRUE</span>,margins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>,<span class="fl">6</span><span class="op">)</span>,useScaled<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>  </span></code></pre></div>
<p><img src="inhibitory_marker_mapping_files/figure-html/Plot%20mFISH%20heatmap%20(scaled)-1.png" width="2304"></p>
<p>A few genes have quite low expression and probably should be omitted
from future experiments. Otherwise, by eye things look reasonable
because there do appear to be distinct expression patterns across
clusters (although it is really hard to tell by eye, and some of the
cells in different blocks seem like they should be grouped
together).</p>
</div>
</div>
<div class="section level2">
<h2 id="quantitative-sanity-check">Quantitative sanity check<a class="anchor" aria-label="anchor" href="#quantitative-sanity-check"></a>
</h2>
<p>So far we have focused on getting the results and trying to determine
agreement with expections based on resulting plots. The results seem
mixed so far. We now want to do what I am calling a quantitative sanity
check, where we compare results obtained by different computational
methods, or between RNA-Seq and mFISH, or taking prior knowledge into
consideration. Ideally we can build a mapping alorithm that adjusts
parameters to try and optimize the results based on priors (or something
to this effect), but for now we want to have quick ways of seeing what
looks right and what looks wrong to help us make adjustments.</p>
<p>First, let’s see whether the proportion of cells identified in
RNA-seq and mFISH agree. Note that we would only expect this to be the
case when we have unbiased surveyed in both modalities, which is not the
case here.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">countFish9</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">fishMouse</span><span class="op">$</span><span class="va">mappingResults</span><span class="op">$</span><span class="va">Class</span><span class="op">)</span></span>
<span><span class="va">countFish9</span> <span class="op">&lt;-</span> <span class="va">countFish9</span><span class="op">[</span><span class="va">countFish9</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">]</span></span>
<span><span class="va">countSeq</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">anno</span><span class="op">$</span><span class="va">primary_type_label</span><span class="op">)</span></span>
<span><span class="va">fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">countFish9</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">countSeq</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countSeq</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countFish9</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span>,main<span class="op">=</span><span class="st">"Sst types"</span>,</span>
<span>  pch<span class="op">=</span><span class="fl">19</span>,col<span class="op">=</span><span class="st">"grey"</span>,ylab<span class="op">=</span><span class="st">"FISH cell count"</span>,xlab<span class="op">=</span><span class="st">"RNA-seq cell count"</span>,xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countSeq</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">*</span><span class="fl">1.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countSeq</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countFish9</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span>,<span class="va">fs</span>,cex<span class="op">=</span><span class="fl">0.7</span>,srt<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="inhibitory_marker_mapping_files/figure-html/type%20wrap-up-1.png" width="768"></p>
<p>There is not great agreement between modalities. What if we wrap up
by class?</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Second, broad class</span></span>
<span><span class="va">val</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fishMouse</span><span class="op">$</span><span class="va">mappingResults</span><span class="op">$</span><span class="va">Class</span>,<span class="va">anno</span><span class="op">$</span><span class="va">primary_type_label</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">val</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">val</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>,<span class="st">" "</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>  <span class="co"># Get the class within interneurons</span></span>
<span><span class="op">}</span></span>
<span><span class="va">countFish</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">val</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">countSeq</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">val</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">val</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="va">val</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countSeq</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countFish</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span>,main<span class="op">=</span><span class="st">"Broad classes"</span>,</span>
<span>  pch<span class="op">=</span><span class="fl">19</span>,col<span class="op">=</span><span class="st">"grey"</span>,ylab<span class="op">=</span><span class="st">"FISH cell count"</span>,xlab<span class="op">=</span><span class="st">"RNA-seq cell count"</span>,</span>
<span>  xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countSeq</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">*</span><span class="fl">1.1</span><span class="op">)</span>,ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countFish</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">*</span><span class="fl">1.1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countSeq</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">countFish</span><span class="op">[</span><span class="va">fs</span><span class="op">]</span><span class="op">)</span>,<span class="va">fs</span>,cex<span class="op">=</span><span class="fl">0.7</span>,srt<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<p><img src="inhibitory_marker_mapping_files/figure-html/broad%20class%20wrap-up-1.png" width="768"></p>
<p>Wrapping up by class, there is much better agreement between
proportions across modalities. Whiel this is not necessarily something
we’d expect to find, it is still useful to note.</p>
<p>One possibility (that I think is likely the case for at least a
subset of cells) is that there is some mapping issues. We can try and
understand why cells are mapping specific ways by plotting expression
levels between RNA-seq and mFISH for each cell type. Let’s do that.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meanFish</span>    <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarizeMatrix.html">summarizeMatrix</a></span><span class="op">(</span><span class="va">fishMouse</span><span class="op">$</span><span class="va">scaleDat</span>,<span class="va">fishMouse</span><span class="op">$</span><span class="va">mappingResults</span><span class="op">$</span><span class="va">Class</span>,summaryFunction <span class="op">=</span> <span class="va">mean</span><span class="op">)</span><span class="op">[</span><span class="va">useGenes</span>,<span class="op">]</span></span>
<span><span class="va">meanFish</span>    <span class="op">&lt;-</span> <span class="va">meanFish</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">meanFish</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">medianExpr</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">medSeq</span>      <span class="op">&lt;-</span> <span class="va">meanExpr</span><span class="op">[</span><span class="va">useGenes</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">meanFish</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">corFS</span>       <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">7</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">ct</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">allClusts</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">medSeq</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">corMR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">medSeq</span><span class="op">[</span>,<span class="va">ct</span><span class="op">]</span>,<span class="va">meanFish</span><span class="op">[</span>,<span class="va">ct</span><span class="op">]</span><span class="op">)</span>,<span class="fl">3</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">medSeq</span><span class="op">[</span>,<span class="va">ct</span><span class="op">]</span>,<span class="va">meanFish</span><span class="op">[</span>,<span class="va">ct</span><span class="op">]</span>,xlab<span class="op">=</span><span class="st">"RNA-seq"</span>,ylab<span class="op">=</span><span class="va">ct</span>,pch<span class="op">=</span><span class="fl">19</span>,col<span class="op">=</span><span class="st">"white"</span>, main<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"R ="</span>,<span class="va">corMR</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="va">medSeq</span><span class="op">[</span>,<span class="va">ct</span><span class="op">]</span>,<span class="va">meanFish</span><span class="op">[</span>,<span class="va">ct</span><span class="op">]</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">medSeq</span><span class="op">)</span>,cex<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span></span>
<span>  <span class="va">corFS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">corFS</span>,<span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">medSeq</span><span class="op">[</span>,<span class="va">ct</span><span class="op">]</span>,<span class="va">meanFish</span><span class="op">[</span>,<span class="va">ct</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">corFS</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">allClusts</span>,<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">medSeq</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="inhibitory_marker_mapping_files/figure-html/mean%20expression%20comparison-1.png" width="1728"></p>
<p>Overall, there is quite good agreement for most, but not all, of
these cell types. Similarly, this type of plot can give us a sense of
which cell types we can be confident in and which ones we can’t,
although the exact relationship between confidence of mapping and
mapping accuracy remains TBD.</p>
<p>Now make a TSNE plot on all genes using the mFISH data. Do the data
cluster by color, and show the points also corresponding to cluster
(abbreviated).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cap</span><span class="op">=</span><span class="fl">10</span></span>
<span><span class="va">fishPlot</span> <span class="op">&lt;-</span> <span class="va">fishMouse</span> </span>
<span><span class="va">fishPlot</span><span class="op">$</span><span class="va">mappingResults</span><span class="op">$</span><span class="va">Broad</span> <span class="op">&lt;-</span> <span class="va">val</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">fishPlot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterCells.html">filterCells</a></span><span class="op">(</span><span class="va">fishPlot</span>,<span class="va">fishPlot</span><span class="op">$</span><span class="va">mappingResults</span><span class="op">$</span><span class="va">Class</span><span class="op">!=</span><span class="st">"none"</span><span class="op">)</span></span>
<span><span class="va">p</span><span class="op">=</span><span class="fu"><a href="../reference/plotTsne.html">plotTsne</a></span><span class="op">(</span><span class="va">fishPlot</span>,main<span class="op">=</span><span class="st">"Mouse cells (scaled)"</span>,colorGroup<span class="op">=</span><span class="st">"Broad"</span>,labelGroup <span class="op">=</span> <span class="st">"Broad"</span>,</span>
<span>           capValue<span class="op">=</span><span class="va">cap</span>, useScaled<span class="op">=</span><span class="cn">TRUE</span>, perplexity <span class="op">=</span> <span class="fl">10</span>, maxNchar<span class="op">=</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="inhibitory_marker_mapping_files/figure-html/Plot%20mFISH%20TSNE%20(scaled)-1.png" width="960"></p>
<p>Overall it seems that most cells map reasonably well by broad class
calls, although it looks like there are a few errors.</p>
<p>Finally, repeat but with the points corresponding to layer.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">=</span><span class="fu"><a href="../reference/plotTsne.html">plotTsne</a></span><span class="op">(</span><span class="va">fishPlot</span>,main<span class="op">=</span><span class="st">"Mouse Sst cells (scaled)"</span>,colorGroup<span class="op">=</span><span class="st">"Class"</span>,labelGroup <span class="op">=</span> <span class="st">"layerData"</span>,</span>
<span>           capValue<span class="op">=</span><span class="va">cap</span>, useScaled<span class="op">=</span><span class="cn">TRUE</span>, perplexity <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="inhibitory_marker_mapping_files/figure-html/Plot%20mFISH%20TSNE%20(scaled,%20color%20by%20layer)-1.png" width="960"></p>
<p>The layer and cell type appear to provide some complementary
information.</p>
<p><strong>It is important to note that these tools are very much in
development and minimally tested, but even so we hope that that are
useful.</strong></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jeremy Miller.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
