<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Marker panel generation • mfishtools</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Marker panel generation">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mfishtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/inhibitory_marker_mapping.html">Mapping mFISH data to reference data set</a>
    </li>
    <li>
      <a href="../articles/inhibitory_marker_selection.html">Marker panel generation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Marker panel generation</h1>
            
      

      <div class="hidden name"><code>inhibitory_marker_selection.Rmd</code></div>

    </div>

    
    
<p>This vignette demonstrates how to generate an optimal marker gene
panel (of length N), and then predicts how well each of the FACs cells
map into the types defined as a measure of panel quality.</p>
<p>The strategy used is a correlation-baseed greedy algorithm, which
aims to minimize the distance between the actual and predicted clusters
(rather than maximizing the fraction correctly mapping to each
cluster).</p>
<div class="section level3">
<h3 id="workspace-set-up">Workspace set-up<a class="anchor" aria-label="anchor" href="#workspace-set-up"></a>
</h3>
<p>Install the necessary packages. In this case we are using data from
<code>tasic2016data</code> and plotting functions from
<code>scrattch.vis</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"AllenInstitute/scrattch.vis"</span><span class="op">)</span>  <span class="co"># For plotting</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"AllenInstitute/tasic2016data"</span><span class="op">)</span> <span class="co"># For our data example</span></span></code></pre></div>
<p>Load libraries.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">mfishtools</span><span class="op">)</span>    <span class="co"># This library!</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/talgalili/gplots" class="external-link">gplots</a></span><span class="op">)</span>        <span class="co"># This is for plotting gene panels only.</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scrattch.vis</span><span class="op">)</span>  <span class="co"># This is for plotting gene panels only.</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">matrixStats</a></span><span class="op">)</span>   <span class="co"># For rowMedians function, which is fast</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">tasic2016data</span><span class="op">)</span> <span class="co"># For the data</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>  <span class="co"># IMPORTANT</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Libraries loaded."</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Libraries loaded."</span></span></code></pre>
<p>Read in the data (in this case we will use the Tasic 2016 data, which
includes ~1800 cells from mouse primary visual cortex).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="va">tasic_2016_anno</span></span>
<span><span class="va">counts</span>      <span class="op">&lt;-</span> <span class="va">tasic_2016_counts</span></span>
<span><span class="va">rpkm</span>        <span class="op">&lt;-</span> <span class="va">tasic_2016_rpkm</span></span>
<span><span class="va">annotations</span> <span class="op">&lt;-</span> <span class="va">annotations</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,<span class="va">annotations</span><span class="op">$</span><span class="va">sample_name</span><span class="op">)</span>,<span class="op">]</span>  <span class="co"># Put them in the correct order</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="data-preparations">Data preparations<a class="anchor" aria-label="anchor" href="#data-preparations"></a>
</h3>
<p>This analysis will only be looking at marker genes for GABAergic
neurons, so we need to only consider cells mapping to GABAergic types.
We also define some convenient cluster info variables here.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clusterType</span> <span class="op">=</span> <span class="va">annotations</span><span class="op">$</span><span class="va">broad_type</span> </span>
<span><span class="va">includeClas</span> <span class="op">=</span> <span class="st">"GABA-ergic Neuron"</span>  <span class="co"># In this analysis, we are only considering interneurons</span></span>
<span><span class="va">excludeClas</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">clusterType</span>,<span class="va">includeClas</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gliaClas</span>    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="va">excludeClas</span>,<span class="st">"Glutamatergic Neuron"</span><span class="op">)</span> </span>
<span><span class="va">kpSamp</span>      <span class="op">=</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">is.element</a></span><span class="op">(</span><span class="va">clusterType</span>,<span class="va">excludeClas</span><span class="op">)</span></span>
<span><span class="va">anno</span>        <span class="op">=</span> <span class="va">annotations</span><span class="op">[</span><span class="va">kpSamp</span>,<span class="op">]</span></span>
<span><span class="va">cl</span>          <span class="op">=</span> <span class="va">annotations</span><span class="op">$</span><span class="va">primary_type_label</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span>   <span class="op">=</span> <span class="va">annotations</span><span class="op">$</span><span class="va">sample_name</span></span>
<span><span class="va">kpClust</span>     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cl</span><span class="op">[</span><span class="va">kpSamp</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gliaClust</span>   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">cl</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">is.element</a></span><span class="op">(</span><span class="va">clusterType</span>,<span class="va">gliaClas</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Convert the data to log2(rpkm). NOTE: we often use counts per million
of introns + exons when performing this analysis. Currently, we don’t
know which method produces more reliable markers. Alternative code for
calculating cpm is commented out below.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">normDat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">rpkm</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#sf      = colSums(counts)/10^6</span></span>
<span><span class="co">#cpms    = t(t(counts)/sf)</span></span>
<span><span class="co">#normDat = log2(cpms+1)</span></span></code></pre></div>
<p>Calculate proportions and medians. These are both needed for gene
filtering and for marker selection.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exprThresh</span> <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="va">medianExpr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span>, <span class="va">cl</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="../reference/rowMedians.html">rowMedians</a></span><span class="op">(</span><span class="va">normDat</span><span class="op">[</span>,<span class="va">x</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">propExpr</span>   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"cbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cl</span><span class="op">)</span>, <span class="va">cl</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">normDat</span><span class="op">[</span>,<span class="va">x</span><span class="op">]</span><span class="op">&gt;</span><span class="va">exprThresh</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">medianExpr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">propExpr</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">normDat</span><span class="op">)</span>  </span></code></pre></div>
</div>
<div class="section level3">
<h3 id="combinatorial-marker-gene-selection">Combinatorial marker gene selection<a class="anchor" aria-label="anchor" href="#combinatorial-marker-gene-selection"></a>
</h3>
<p>This section is where gene selection happens. There are two steps:
(1) gene filtering and (2) marker selection, both of which are described
below.</p>
<p>We first want to define some gene filters prior to running gene
selection. Note that this filtering occurs prior to gene selection and
does not occur during the selection process. To see details about all of
the specific filters, see the code block below or use
<code><a href="../reference/filterPanelGenes.html">?filterPanelGenes</a></code>. Overall, the goal of this section is to
exclude genes that won’t work properly with the specific spatial
transcriptomics method desired because their expression is too low or
too high, or they are too short. It also removes genes with too much
off-target expression, genes that are expressed in too many cell types,
and genes that are potentially not of interest because they are
unannotated, or on a sex chromosome, or are don’t work for any other
reason. It also takes the most binary genes as the possible selection
space in order to try and avoid genes whose only cell type differences
are in magnitude. In this case, we will use a total of 250 binary
genes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">startingGenePanel</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Gad1"</span>,<span class="st">"Slc32a1"</span>,<span class="st">"Pvalb"</span>,<span class="st">"Sst"</span>,<span class="st">"Vip"</span><span class="op">)</span></span>
<span><span class="va">runGenes</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">runGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filterPanelGenes.html">filterPanelGenes</a></span><span class="op">(</span></span>
<span>  summaryExpr <span class="op">=</span> <span class="fl">2</span><span class="op">^</span><span class="va">medianExpr</span><span class="op">-</span><span class="fl">1</span>,  <span class="co"># medians (could also try means); We enter linear values to match the linear limits below</span></span>
<span>  propExpr    <span class="op">=</span> <span class="va">propExpr</span>,    <span class="co"># proportions</span></span>
<span>  onClusters  <span class="op">=</span> <span class="va">kpClust</span>,     <span class="co"># clusters of interest for gene panel</span></span>
<span>  offClusters <span class="op">=</span> <span class="va">gliaClust</span>,   <span class="co"># clusters to exclude expression</span></span>
<span>  geneLengths <span class="op">=</span> <span class="cn">NULL</span>,        <span class="co"># vector of gene lengths (not included here)</span></span>
<span>  startingGenes  <span class="op">=</span> <span class="va">startingGenePanel</span>,  <span class="co"># Starting genes (from above)</span></span>
<span>  numBinaryGenes <span class="op">=</span> <span class="fl">250</span>,      <span class="co"># Number of binary genes (explained below)</span></span>
<span>  minOn     <span class="op">=</span> <span class="fl">10</span>,   <span class="co"># Minimum required expression in highest expressing cell type</span></span>
<span>  maxOn     <span class="op">=</span> <span class="fl">500</span>,  <span class="co"># Maximum allowed expression</span></span>
<span>  maxOff    <span class="op">=</span> <span class="fl">50</span>,   <span class="co"># Maximum allowed expression in off types (e.g., aviod glial expression)</span></span>
<span>  minLength <span class="op">=</span> <span class="fl">960</span>,  <span class="co"># Minimum gene length (to allow probe design; ignored in this case)</span></span>
<span>  fractionOnClusters <span class="op">=</span> <span class="fl">0.5</span>,  <span class="co"># Max fraction of on clusters (described above)</span></span>
<span>  excludeGenes    <span class="op">=</span> <span class="cn">NULL</span>,    <span class="co"># Genes to exclude.  Often sex chromosome or mitochondrial genes would be input here.</span></span>
<span>  excludeFamilies <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"LOC"</span>,<span class="st">"Fam"</span>,<span class="st">"RIK"</span>,<span class="st">"RPS"</span>,<span class="st">"RPL"</span>,<span class="st">"\\-"</span>,<span class="st">"Gm"</span>,<span class="st">"Rnf"</span>,<span class="st">"BC0"</span><span class="op">)</span><span class="op">)</span> <span class="co"># Avoid LOC markers, in this case</span></span></code></pre></div>
<pre><code><span><span class="co">## 1278 total genes pass constraints prior to binary score calculation.</span></span></code></pre>
<p>The second step is our marker panel selection. This strategy uses a
greedy algorithm to iteratively add the “best” gene to the existing
panel until the panel reaches a certain size. Specifically, each cell in
the reference data set is correlated with each cluster median using the
existing marker gene set, and the most highly correlated cluster is
compared with the originally assigned cluster for each cell. By default
the algorithm tries to optimize the fraction of cells correctly mapping
to each cluster by using this correlation-based mapping. An alternative
strategy that is particularly useful for smaller gene panels is to use a
weighting strategy to penalize cells that map to a distant cluster more
than cells that map to a nearby cluster. To do this we iteratively
choose genes from the starting gene panel such that the addition of each
gene minimizes the correlation distance between clusters.</p>
<p>To do this we first determine the cluster distance based on
correlation of top binary genes, which is what is done in this code
block.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">corDist</span>         <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">clusterDistance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu">corDist</span><span class="op">(</span><span class="va">medianExpr</span><span class="op">[</span><span class="va">runGenes</span>,<span class="va">kpClust</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">clusterDistance</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 23 23</span></span></code></pre>
<p>This step constructs the gene panel, as described above. Once again,
use <code><a href="../reference/buildMappingBasedMarkerPanel.html">?buildMappingBasedMarkerPanel</a></code> and see the code block
below for details on the parameters, but there are a few key options.
First, we find it useful to subsample cells from each cluster which
decreases the time for the algorithm to run and also more evenly weights
the clusters for gene selection.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fishPanel</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildMappingBasedMarkerPanel.html">buildMappingBasedMarkerPanel</a></span><span class="op">(</span></span>
<span>  mapDat        <span class="op">=</span> <span class="va">normDat</span><span class="op">[</span><span class="va">runGenes</span>,<span class="va">kpSamp</span><span class="op">]</span>,     <span class="co"># Data for optimization</span></span>
<span>  medianDat     <span class="op">=</span> <span class="va">medianExpr</span><span class="op">[</span><span class="va">runGenes</span>,<span class="va">kpClust</span><span class="op">]</span>, <span class="co"># Median expression levels of relevant genes in relevant clusters</span></span>
<span>  clustersF     <span class="op">=</span> <span class="va">cl</span><span class="op">[</span><span class="va">kpSamp</span><span class="op">]</span>,                   <span class="co"># Vector of cluster assignments</span></span>
<span>  panelSize     <span class="op">=</span> <span class="fl">30</span>,                           <span class="co"># Final panel size</span></span>
<span>  currentPanel  <span class="op">=</span> <span class="va">startingGenePanel</span>,            <span class="co"># Starting gene panel</span></span>
<span>  subSamp       <span class="op">=</span> <span class="fl">15</span>,                           <span class="co"># Maximum number of cells per cluster to include in analysis (20-50 is usually best)</span></span>
<span>  optimize      <span class="op">=</span> <span class="st">"CorrelationDistance"</span>,        <span class="co"># CorrelationDistance maximizes the cluster distance as described</span></span>
<span>  clusterDistance <span class="op">=</span> <span class="va">clusterDistance</span>,            <span class="co"># Cluster distance matrix</span></span>
<span>  percentSubset <span class="op">=</span> <span class="fl">50</span>                            <span class="co"># Only consider a certain percent of genes each iteration to speed up calculations (in most cases this is not recommeded)</span></span>
<span><span class="op">)</span>       </span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## [1] "Added Necab1 with average cluster distance 0.248 [ 5 ]."</span></span>
<span><span class="co">## [1] "Added Col25a1 with average cluster distance 0.212 [ 6 ]."</span></span>
<span><span class="co">## [1] "Added Thsd7a with average cluster distance 0.184 [ 7 ]."</span></span>
<span><span class="co">## [1] "Added Adarb2 with average cluster distance 0.166 [ 8 ]."</span></span>
<span><span class="co">## [1] "Added Adra1b with average cluster distance 0.15 [ 9 ]."</span></span>
<span><span class="co">## [1] "Added Crh with average cluster distance 0.138 [ 10 ]."</span></span>
<span><span class="co">## [1] "Added Sstr1 with average cluster distance 0.126 [ 11 ]."</span></span>
<span><span class="co">## [1] "Added Trp53i11 with average cluster distance 0.121 [ 12 ]."</span></span>
<span><span class="co">## [1] "Added Cpne2 with average cluster distance 0.111 [ 13 ]."</span></span>
<span><span class="co">## [1] "Added Rgs12 with average cluster distance 0.104 [ 14 ]."</span></span>
<span><span class="co">## [1] "Added Npy2r with average cluster distance 0.0989 [ 15 ]."</span></span>
<span><span class="co">## [1] "Added Cbln4 with average cluster distance 0.0926 [ 16 ]."</span></span>
<span><span class="co">## [1] "Added Grpr with average cluster distance 0.0907 [ 17 ]."</span></span>
<span><span class="co">## [1] "Added Crispld2 with average cluster distance 0.0874 [ 18 ]."</span></span>
<span><span class="co">## [1] "Added Cacna2d1 with average cluster distance 0.0826 [ 19 ]."</span></span>
<span><span class="co">## [1] "Added Cplx3 with average cluster distance 0.078 [ 20 ]."</span></span>
<span><span class="co">## [1] "Added Sorcs1 with average cluster distance 0.0746 [ 21 ]."</span></span>
<span><span class="co">## [1] "Added Npas1 with average cluster distance 0.0705 [ 22 ]."</span></span>
<span><span class="co">## [1] "Added Tacstd2 with average cluster distance 0.0694 [ 23 ]."</span></span>
<span><span class="co">## [1] "Added Ttc39b with average cluster distance 0.0678 [ 24 ]."</span></span>
<span><span class="co">## [1] "Added Gpc3 with average cluster distance 0.0675 [ 25 ]."</span></span>
<span><span class="co">## [1] "Added Nek7 with average cluster distance 0.066 [ 26 ]."</span></span>
<span><span class="co">## [1] "Added Cacna1e with average cluster distance 0.0629 [ 27 ]."</span></span>
<span><span class="co">## [1] "Added Olfm3 with average cluster distance 0.0604 [ 28 ]."</span></span>
<span><span class="co">## [1] "Added Ano3 with average cluster distance 0.0576 [ 29 ]."</span></span></code></pre>
<p>This is the panel!</p>
</div>
<div class="section level3">
<h3 id="assess-panel-quality">Assess panel quality<a class="anchor" aria-label="anchor" href="#assess-panel-quality"></a>
</h3>
<p>First, let’s plot the panel in the context of the clusters we care
about. This can be done using the <code>scrattch.vis</code> library.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plotGenes</span> <span class="op">&lt;-</span> <span class="va">fishPanel</span></span>
<span><span class="va">plotData</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>sample_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">rpkm</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">rpkm</span><span class="op">[</span><span class="va">plotGenes</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">clid_inh</span>  <span class="op">&lt;-</span> <span class="fl">1</span><span class="op">:</span><span class="fl">23</span>  <span class="co"># Cluster IDs for inhibitory clusters</span></span>
<span></span>
<span><span class="co"># violin plot example.  Could be swapped with fire, dot, bar, box plot, heatmap, Quasirandom, etc.</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scrattch.vis/man/sample_fire_plot.html" class="external-link">sample_fire_plot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">plotData</span>, anno <span class="op">=</span> <span class="va">annotations</span>, genes <span class="op">=</span> <span class="va">plotGenes</span>, grouping <span class="op">=</span> <span class="st">"primary_type"</span>, </span>
<span>                 log_scale<span class="op">=</span><span class="cn">TRUE</span>, max_width<span class="op">=</span><span class="fl">15</span>, label_height <span class="op">=</span> <span class="fl">8</span>, group_order <span class="op">=</span> <span class="va">clid_inh</span><span class="op">)</span></span></code></pre></div>
<p><img src="inhibitory_marker_selection_files/figure-html/panel%20gene%20plot%20in%20subset-1.png" width="672"></p>
<p>The first half of this panel looks like most of the genes have fairly
distinct and binarized patterning, while many of the genes in the latter
half of the panel appear to be showing redundant information, suggesting
that after a certain point this strategy becomes less than optimal in a
practical sense.</p>
<p>How do they look across ALL cell types (realizing that we don’t
expect to see all these types in the tissue we care about)?</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scrattch.vis/man/sample_fire_plot.html" class="external-link">sample_fire_plot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">plotData</span>, anno <span class="op">=</span> <span class="va">annotations</span>, genes <span class="op">=</span> <span class="va">plotGenes</span>, grouping <span class="op">=</span> <span class="st">"primary_type"</span>, </span>
<span>                 log_scale<span class="op">=</span><span class="cn">TRUE</span>, max_width<span class="op">=</span><span class="fl">15</span>, label_height <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></span></code></pre></div>
<p><img src="inhibitory_marker_selection_files/figure-html/panel%20gene%20plot%20across%20all%20types-1.png" width="1344"></p>
<p>We get a little bit of excitatory and glial cell separation for free,
but it’s not great (as expected).</p>
<p>What fraction of cells are correctly mapped to leaf nodes? Note that
we don’t necessarily expect this number to be high. Also note that this
is using all of the above genes, so will actually be higher than we
would get with any given 9-gene panel.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">assignedCluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/getTopMatch.html">getTopMatch</a></span><span class="op">(</span><span class="fu"><a href="../reference/corTreeMapping.html">corTreeMapping</a></span><span class="op">(</span>mapDat <span class="op">=</span> <span class="va">normDat</span><span class="op">[</span><span class="va">runGenes</span>,<span class="va">kpSamp</span><span class="op">]</span>, </span>
<span>                   medianDat<span class="op">=</span><span class="va">medianExpr</span><span class="op">[</span><span class="va">runGenes</span>,<span class="va">kpClust</span><span class="op">]</span>, genesToMap<span class="op">=</span><span class="va">fishPanel</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Percent correctly mapped: "</span>,<span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="fl">100</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">assignedCluster</span><span class="op">$</span><span class="va">TopLeaf</span><span class="op">)</span><span class="op">==</span><span class="va">cl</span><span class="op">[</span><span class="va">kpSamp</span><span class="op">]</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,<span class="fl">3</span><span class="op">)</span>,<span class="st">"%"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Percent correctly mapped: 68.5%"</span></span></code></pre>
<p>Around ~72% of the cells are correctly mapped with this panel, which
is reasonable, but less than ideal. How does the plot look for fewer
genes?</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fractionCorrectWithGenes.html">fractionCorrectWithGenes</a></span><span class="op">(</span><span class="va">fishPanel</span>,<span class="va">normDat</span><span class="op">[</span>,<span class="va">kpSamp</span><span class="op">]</span>,<span class="va">medianExpr</span><span class="op">[</span><span class="va">runGenes</span>,<span class="va">kpClust</span><span class="op">]</span>,<span class="va">cl</span><span class="op">[</span><span class="va">kpSamp</span><span class="op">]</span>,</span>
<span>                         main<span class="op">=</span><span class="st">"Mapping quality for different numbers of included genes"</span>,return<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="inhibitory_marker_selection_files/figure-html/plot%20fraction%20correctly%20mapped-1.png" width="864"></p>
<p>This plot suggests that with 30 genes we are continuing to see
improvement by adding new genes; however, by approximately a 20 gene
panel the level of improvement decreases dramatically for each gene
added. <strong>Later releases of this library will include additional
strategies for optimizing the panel beyond these initial 20 or so
genes.</strong></p>
<p>Finally, as an overview, we create a confusion matrix based on the
top leaf assignments. This will let us address which pairs of clusters
are the most confused. Are they adjacent/nearby on the tree? Note that
the colors are distorted to highlight confusion in the tree.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">membConfusionProp</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getConfusionMatrix.html">getConfusionMatrix</a></span><span class="op">(</span><span class="va">cl</span><span class="op">[</span><span class="va">kpSamp</span><span class="op">]</span>,<span class="va">assignedCluster</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,<span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">clOrd</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">annotations</span><span class="op">$</span><span class="va">primary_type_label</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">clid_inh</span>,<span class="va">annotations</span><span class="op">$</span><span class="va">primary_type_id</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>  <span class="co"># Cluster order</span></span>
<span><span class="fu"><a href="https://talgalili.github.io/gplots/reference/heatmap.2.html" class="external-link">heatmap.2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="va">membConfusionProp</span>,<span class="fl">0.25</span><span class="op">)</span><span class="op">[</span><span class="va">clOrd</span>,<span class="va">clOrd</span><span class="op">]</span>,Rowv<span class="op">=</span><span class="cn">FALSE</span>,Colv<span class="op">=</span><span class="cn">FALSE</span>,trace<span class="op">=</span><span class="st">"none"</span>,dendrogram<span class="op">=</span><span class="st">"none"</span>,</span>
<span>          margins<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">16</span>,<span class="fl">16</span><span class="op">)</span>,main<span class="op">=</span><span class="st">"Confusion Matrix"</span><span class="op">)</span></span></code></pre></div>
<p><img src="inhibitory_marker_selection_files/figure-html/cluster%20confusion-1.png" width="768"></p>
<p>Most of the errors are nearby on the diagonal, which is what we were
optimizing for using the clusterDistance strategy.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Jeremy Miller.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
