<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Package index â€¢ mfishtools</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mfishtools</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/inhibitory_marker_mapping.html">Mapping mFISH data to reference data set</a>
    </li>
    <li>
      <a href="../articles/inhibitory_marker_selection.html">Marker panel generation</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="buildMappingBasedMarkerPanel.html">buildMappingBasedMarkerPanel()</a></code> </p>
        </td>
        <td><p>Greedy algorithm for building marker gene panel</p></td>
      </tr><tr><td>
          <p><code><a href="buildPanel_oneCluster.html">buildPanel_oneCluster()</a></code> </p>
        </td>
        <td><p>Build panel for one cluster (beta)</p></td>
      </tr><tr><td>
          <p><code><a href="buildQualityTable.html">buildQualityTable()</a></code> </p>
        </td>
        <td><p>Correct mapping at different tree heights</p></td>
      </tr><tr><td>
          <p><code><a href="buildTreeFromGenePanel.html">buildTreeFromGenePanel()</a></code> </p>
        </td>
        <td><p>Build and plot dendrogram from gene panel</p></td>
      </tr><tr><td>
          <p><code><a href="cellToClusterMapping_byCor.html">cellToClusterMapping_byCor()</a></code> </p>
        </td>
        <td><p>Return top mapped correlation-based cluster and confidence</p></td>
      </tr><tr><td>
          <p><code><a href="cellToClusterMapping_byRank.html">cellToClusterMapping_byRank()</a></code> </p>
        </td>
        <td><p>Cell-based cluster mapping</p></td>
      </tr><tr><td>
          <p><code><a href="corTreeMapping.html">corTreeMapping()</a></code> </p>
        </td>
        <td><p>Correlation-based cluster mapping</p></td>
      </tr><tr><td>
          <p><code><a href="corTreeMapping_withFilter.html">corTreeMapping_withFilter()</a></code> </p>
        </td>
        <td><p>Correlation between nodes and leafs (deprecated)</p></td>
      </tr><tr><td>
          <p><code><a href="distTreeMapping.html">distTreeMapping()</a></code> </p>
        </td>
        <td><p>(Euclidean) distance mapping</p></td>
      </tr><tr><td>
          <p><code><a href="filterByClass.html">filterByClass()</a></code> </p>
        </td>
        <td><p>Filter by meta-data</p></td>
      </tr><tr><td>
          <p><code><a href="filterCells.html">filterCells()</a></code> </p>
        </td>
        <td><p>Filter (subset) fishScaleAndMap object</p></td>
      </tr><tr><td>
          <p><code><a href="filterPanelGenes.html">filterPanelGenes()</a></code> </p>
        </td>
        <td><p>Filter genes for spatial transcriptomics panel</p></td>
      </tr><tr><td>
          <p><code><a href="fishScaleAndMap.html">fishScaleAndMap()</a></code> </p>
        </td>
        <td><p>Scale mFISH data and map to RNA-seq reference</p></td>
      </tr><tr><td>
          <p><code><a href="fractionCorrectPerNode.html">fractionCorrectPerNode()</a></code> </p>
        </td>
        <td><p>Fraction of correct calls per node</p></td>
      </tr><tr><td>
          <p><code><a href="fractionCorrectWithGenes.html">fractionCorrectWithGenes()</a></code> </p>
        </td>
        <td><p>Fraction of cells correctly assigned</p></td>
      </tr><tr><td>
          <p><code><a href="generateMultipleCellReferenceSet.html">generateMultipleCellReferenceSet()</a></code> </p>
        </td>
        <td><p>Generate reference set of pseudo-cells</p></td>
      </tr><tr><td>
          <p><code><a href="getBetaScore.html">getBetaScore()</a></code> </p>
        </td>
        <td><p>Get binary (aka beta) score</p></td>
      </tr><tr><td>
          <p><code><a href="getBranchList.html">getBranchList()</a></code> </p>
        </td>
        <td><p>Branch list</p></td>
      </tr><tr><td>
          <p><code><a href="getConfusionMatrix.html">getConfusionMatrix()</a></code> </p>
        </td>
        <td><p>Confusion matrix</p></td>
      </tr><tr><td>
          <p><code><a href="getDend.html">getDend()</a></code> </p>
        </td>
        <td><p>Build a dendrogram from gene panel</p></td>
      </tr><tr><td>
          <p><code><a href="getNodeHeight.html">getNodeHeight()</a></code> </p>
        </td>
        <td><p>Get node height</p></td>
      </tr><tr><td>
          <p><code><a href="getTopMatch.html">getTopMatch()</a></code> </p>
        </td>
        <td><p>Get top leaf match</p></td>
      </tr><tr><td>
          <p><code><a href="get_subtree_label.html">get_subtree_label()</a></code> </p>
        </td>
        <td><p>Gets subtree labels for lca function.</p></td>
      </tr><tr><td>
          <p><code><a href="labelDend.html">labelDend()</a></code> </p>
        </td>
        <td><p>Label dendrogram nodes</p></td>
      </tr><tr><td>
          <p><code><a href="layerFraction.html">layerFraction()</a></code> </p>
        </td>
        <td><p>Layer weights per cell</p></td>
      </tr><tr><td>
          <p><code><a href="layerScale.html">layerScale()</a></code> </p>
        </td>
        <td><p>Fraction of cells per layer</p></td>
      </tr><tr><td>
          <p><code><a href="lca.html">lca()</a></code> </p>
        </td>
        <td><p>Get lowest common ancestor (defined cluster pairs)</p></td>
      </tr><tr><td>
          <p><code><a href="leafToNodeMedians.html">leafToNodeMedians()</a></code> </p>
        </td>
        <td><p>Return mean node expression</p></td>
      </tr><tr><td>
          <p><code><a href="makeLCAtable.html">makeLCAtable()</a></code> </p>
        </td>
        <td><p>Get lowest common ancestor (all cluster pairs in tree)</p></td>
      </tr><tr><td>
          <p><code><a href="map_dend.html">map_dend()</a></code> </p>
        </td>
        <td><p>Tree-based mapping</p></td>
      </tr><tr><td>
          <p><code><a href="mergeFish.html">mergeFish()</a></code> </p>
        </td>
        <td><p>Merge two fishScaleAndMap objects</p></td>
      </tr><tr><td>
          <p><code><a href="mfishtools.html">mfishtools-package</a></code> <code><a href="mfishtools.html">mfishtools</a></code> </p>
        </td>
        <td><p>mfishtools: Building Gene Sets and Mapping mFISH Data.</p></td>
      </tr><tr><td>
          <p><code><a href="outputTopConfused.html">outputTopConfused()</a></code> </p>
        </td>
        <td><p>Table of confused clusters</p></td>
      </tr><tr><td>
          <p><code><a href="plotConfusionVsConfidence.html">plotConfusionVsConfidence()</a></code> </p>
        </td>
        <td><p>Confusion plot vs. confidence</p></td>
      </tr><tr><td>
          <p><code><a href="plotCorrectWithGenes.html">plotCorrectWithGenes()</a></code> </p>
        </td>
        <td><p>Plot fraction correct</p></td>
      </tr><tr><td>
          <p><code><a href="plotDistributions.html">plotDistributions()</a></code> </p>
        </td>
        <td><p>Plot distributions</p></td>
      </tr><tr><td>
          <p><code><a href="plotHeatmap.html">plotHeatmap()</a></code> </p>
        </td>
        <td><p>Plot heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotNodes.html">plotNodes()</a></code> </p>
        </td>
        <td><p>Plot dendrogram</p></td>
      </tr><tr><td>
          <p><code><a href="plotTsne.html">plotTsne()</a></code> </p>
        </td>
        <td><p>Plot TSNE</p></td>
      </tr><tr><td>
          <p><code><a href="possibleClustersByPriors.html">possibleClustersByPriors()</a></code> </p>
        </td>
        <td><p>Filter possible cluster calls using priors</p></td>
      </tr><tr><td>
          <p><code><a href="quantileTruncate.html">quantileTruncate()</a></code> </p>
        </td>
        <td><p>Quantile normalize, truncate, and scale</p></td>
      </tr><tr><td>
          <p><code><a href="resolve_cl.html">resolve_cl()</a></code> </p>
        </td>
        <td><p>Tree-based mapping (internal)</p></td>
      </tr><tr><td>
          <p><code><a href="rfTreeMapping.html">rfTreeMapping()</a></code> </p>
        </td>
        <td><p>Tree-based mapping</p></td>
      </tr><tr><td>
          <p><code><a href="rotateXY.html">rotateXY()</a></code> </p>
        </td>
        <td><p>Rotate coordinates</p></td>
      </tr><tr><td>
          <p><code><a href="rowMedians.html">rowMedians()</a></code> </p>
        </td>
        <td><p>Calculate row medians</p></td>
      </tr><tr><td>
          <p><code><a href="smartLayerAllocation.html">smartLayerAllocation()</a></code> </p>
        </td>
        <td><p>Layer weights per cell</p></td>
      </tr><tr><td>
          <p><code><a href="subsampleCells.html">subsampleCells()</a></code> </p>
        </td>
        <td><p>Subsample cells</p></td>
      </tr><tr><td>
          <p><code><a href="summarizeMatrix.html">summarizeMatrix()</a></code> </p>
        </td>
        <td><p>Summarize matrix</p></td>
      </tr><tr><td>
          <p><code><a href="update_mfishtools.html">update_mfishtools()</a></code> </p>
        </td>
        <td><p>Update the mfishtools library</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jeremy Miller.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

